reference.go

```diff
@@ -243,18 +243,21 @@ func ParseNamed(s string) (Named, error) {
 // WithName returns a named object representing the given string. If the input
 // is invalid ErrReferenceInvalidFormat will be returned.
 func WithName(name string) (Named, error) {
-	if len(name) > NameTotalLengthMax {
-		return nil, ErrNameTooLong
-	}
-
 	match := anchoredNameRegexp.FindStringSubmatch(name)
 	if match == nil || len(match) != 3 {
 		return nil, ErrReferenceInvalidFormat
 	}
-	return repository{
+
+	repo := repository{
 		domain: match[1],
 		path:   match[2],
-	}, nil
+	}
+
+	if len(repo.path) > RepositoryNameTotalLengthMax {
```

Comment:
Perhaps validate `match[2]`, so that we can keep the return more atomic (`return repository{ ...}`), instead of constructing the `repository` and discarding it in case there's a failure.